libs/langchain/langchain/vectorstores/pinecone.py

```diff
@@ -93,34 +99,60 @@ def add_texts(
         ids: Optional[List[str]] = None,
         namespace: Optional[str] = None,
         batch_size: int = 32,
+        embedding_chunk_size: int = 1000,
         **kwargs: Any,
     ) -> List[str]:
         """Run more texts through the embeddings and add to the vectorstore.
 
+        Upsert optimization is done by chunking the embeddings and upserting them.
+        This is done to avoid memory issues and optimize using HTTP based embeddings.
+        For OpenAI embeddings, use pool_threads>4 when constructing the pinecone.Index,
+        embedding_chunk_size>1000 and batch_size~64 for best performance.
         Args:
             texts: Iterable of strings to add to the vectorstore.
             metadatas: Optional list of metadatas associated with the texts.
             ids: Optional list of ids to associate with the texts.
             namespace: Optional pinecone namespace to add the texts to.
+            batch_size: Batch size to use when adding the texts to the vectorstore.
+            embedding_chunk_size: Chunk size to use when embedding the texts.
 
         Returns:
             List of ids from adding the texts into the vectorstore.
 
         """
         if namespace is None:
             namespace = self._namespace
-        # Embed and create the documents
-        docs = []
+
+        texts = list(texts)
         ids = ids or [str(uuid.uuid4()) for _ in texts]
-        embeddings = self._embed_documents(texts)
-        for i, (text, embedding) in enumerate(zip(texts, embeddings)):
-            metadata = metadatas[i] if metadatas else {}
+        metadatas = metadatas or [{} for _ in texts]
+        for index, text in enumerate(texts):
+            metadata = metadatas[index]
             metadata[self._text_key] = text
-            docs.append((ids[i], embedding, metadata))
-        # upsert to Pinecone
-        self._index.upsert(
-            vectors=docs, namespace=namespace, batch_size=batch_size, **kwargs
-        )
+
+        # For loops to avoid memory issues and optimize when using HTTP based embeddings
+        # The first loop runs the embeddings, it benefits when using OpenAI embeddings
+        # The second loops runs the pinecone upsert asynchoronously.
+        for i in range(0, len(texts), embedding_chunk_size):
+            chunk_texts = texts[i : i + embedding_chunk_size]
+            chunk_ids = ids[i : i + embedding_chunk_size]
+            chunk_metadatas = metadatas[i : i + embedding_chunk_size]
+            embeddings = self._embed_documents(chunk_texts)
+            async_res = [
+                # TODO: add tenacity retry logic
```

Comment:
Is this TODO still outstanding?

Reply:
With `thread_pools` and `batch_size` we have a control to avoid limits and throttling, not sure do we still need to add retry mechanism? I thought this might be better to implement it inside pinecone python-client (if there isn't anything like that), not necessary inside langchain.